File: lintcode/919.meeting-rooms-ii.cpp
#include <iostream>
#include <queue>
#include <stack>
#include <vector>

using namespace std;

#ifdef DEBUG
class Interval {
public:
  int start, end;
  Interval(int start, int end) : start{start}, end{end} {}
};
#endif

class Solution {
public:
  /**
   * @param intervals: an array of meeting time intervals
   * @return: the minimum number of conference rooms required
   */
  int minMeetingRooms(vector<Interval> &intervals) {
    sort(intervals.begin(), intervals.end(),
         [](const Interval &i1, const Interval &i2) {
           return i1.start < i2.start;
         });

    auto cmp = [](const Interval &i1, const Interval &i2) {
      return i1.end > i2.end;
    };

    priority_queue<Interval, vector<Interval>, decltype(cmp)> q{cmp};
    for (auto &i : intervals) {
      cout << i.start << " " << i.end << endl;
      if (!q.empty() && q.top().end <= i.start) {
        q.pop();
      }
      q.push(i);
    }
    return q.size();
  }
};

#ifdef DEBUG
int main() {
  Interval i1{0, 30};
  Interval i2{5, 10};
  Interval i3{15, 20};

  vector<Interval> intervals{i1, i2, i3};

  Solution sol;
  cout << sol.minMeetingRooms(intervals) << endl;
}
#endif

File: lintcode/919.meeting-rooms-ii.py
"""
class Interval(object):
    def __init__(self, start, end):
        self.start = start
        self.end = end
"""


class Solution:
    """
    @param intervals: an array of meeting time intervals
    @return: the minimum number of conference rooms required
    """

    def minMeetingRooms(self, intervals):
        intervals.sort(key=lambda i: i.start)

        import heapq
        heap = []

        for interval in intervals:
            if len(heap) > 0 and heap[0] <= interval.start:
                heapq.heappop(heap)
            heapq.heappush(heap, interval.end)
        return len(heap)


if __name__ == '__main__':
    i1 = Interval(0, 30)
    i2 = Interval(5, 10)
    i3 = Interval(15, 20)

    s = Solution()

    print(s.minMeetingRooms([i1, i2, i3]))
